---
title: "Leukemia DE analysis for Edwin"
author: "Gosia Nowicka"
date: "20 Mar 2015"
output: 
  html_document:
    toc: true
---

# Data and experiment description

Affymetrix Mouse Gene 2.0 ST arrays

# Prepare data

```{r setup, echo = FALSE}
opts_knit$set(root.dir = "/home/Shared/data/array/Microarray_Edwin")
opts_chunk$set(message=FALSE, warning = FALSE, cache = TRUE, size="small", tidy=TRUE, tidy.opts=list(blank=FALSE, width.cutoff=70), dev = "pdf", cache.path = '/home/Shared/data/array/Microarray_Edwin/Analysis_Mouse_cache/', fig.path = '/home/Shared/data/array/Microarray_Edwin/Analysis_Mouse_figure/', fig.width = 7, fig.height = 7)
options(width=100)
```

Load targets table with information about samples.

```{r}
targets <- read.table(file.path("metadata", "targets.xls"), header = TRUE, sep = "\t", comment.char = "", as.is = TRUE)

targets

```

Load Affy CEL data using `oligo` package. Plot the distributions of raw probe expression in all samples.

```{r}
library(oligo)
library(pd.mogene.2.0.st)

ff <- as.character(targets$FileName)

x <- oligo::read.celfiles(filenames = ff)

par(mar = c(12, 4, 4, 2) + 0.1) # c(bottom, left, top, right), default = c(5, 4, 4, 2) + 0.1
boxplot(x, las = 2, col = targets$colors, names = targets$labels, las=2)

par(mar = c(5, 4, 4, 2) + 0.1)
hist(x, col = targets$colors, lwd = 2)
legend("topright", legend = targets$labels, col =  targets$colors, lty = 1, lwd = 2, cex = 0.8)

```

# Normalize data with probe-level model

Run `fitProbeLevelModel` and create NUSE and RLE plots (https://genevestigator.com/userdocs/manual/qc.html).

```{r}
fitplm <- oligo::fitProbeLevelModel(x)

par(mar = c(12, 4, 4, 2) + 0.1) # c(bottom, left, top, right), default = c(5, 4, 4, 2) + 0.1
oligo::NUSE(fitplm, col = targets$colors, names = targets$labels, las=2)

par(mar = c(12, 4, 4, 2) + 0.1) # c(bottom, left, top, right), default = c(5, 4, 4, 2) + 0.1
oligo::RLE(fitplm, col = targets$colors, names = targets$labels, las=2)

```

# Normalize data with RMA method

```{r}
eset.org <- eset <- oligo::rma(x) ## Is the expression in log2 scale?

par(mar = c(12, 4, 4, 2) + 0.1) # c(bottom, left, top, right), default = c(5, 4, 4, 2) + 0.1
boxplot(eset, las = 2, col = targets$colors, names = targets$labels)

par(mar = c(5, 4, 4, 2) + 0.1)
hist(eset, col = targets$colors, lwd = 2)
legend("topright", legend = targets$labels, col =  targets$colors, lty = 1, lwd = 2, cex = 0.8)

library(limma)
labels <- paste0(ifelse(is.na(targets$ctrlRep), "", paste0(targets$ctrlRep, " ")), targets$labels)

### plot MDS for all samples
mds.dendo <- mds <- plotMDS(eset, top=1000, col = targets$colors, labels = labels, cex = 1.2)
### plot only points
plot(mds$x, mds$y, pch = 16, cex = 1.5, col = targets$colors)


library(ggplot2)
library(ggdendro)

d <- mds.dendo$distance.matrix
rownames(d) <- labels
hc <- hclust(as.dist(d), method = "single")
ggdendrogram(hc, theme_dendro = TRUE) + theme(axis.text.x = element_text(colour = targets$colors[hc$order], size = 13)) 


```

```{r}
####### Do NOT keep the HeLa sample for the rest of the analysis

keepSAMPS <- targets$labels != "control_HeLa"

eset.org <- eset <- eset[, keepSAMPS]
targets.org <- targets <- targets[keepSAMPS, ]

```




# Get NetAffx annotation of probesets

```{r}
####### NetAffx Annotation with getNetAffx()

infoNetAffx <- pData(getNetAffx(eset, "transcript"))
colnames(infoNetAffx)

```



# Filtering

Keep only the main probes.


```{r}
table(infoNetAffx$category, useNA = "always")

keepMAIN <- infoNetAffx$category == "main"

eset.main <- eset[keepMAIN, ]

```



Keep only the probes that come from chr 1-19, X and Y.

```{r}
keepCHR <- featureNames(eset.main) %in% rownames(infoNetAffx)[which(infoNetAffx$seqname %in% paste0("chr", c(1:19, "Y", "X")), useNames = TRUE)]

table(keepCHR)

eset.main <- eset.main[keepCHR, ]

```


# Get annotation from formated files from Affy website

http://www.affymetrix.com/analysis/index.affx

```{r}
library(limma)

### files that are formated for easy load 
annof <- list.files("NetAffx", pattern = ".tsv", full.names = TRUE)
annof

### use gene_list

allLines <- readLines(annof[grepl("gene_list", annof)], n=-1)

gl <- data.frame(strsplit2(allLines, "\t"), stringsAsFactors = FALSE)
colnames(gl) <- gsub(pattern = " ", replacement = "" ,gl[1,])
gl <- gl[-1,]
rownames(gl) <- gl$TranscriptClusterID

colnames(gl)

annot <- gl[ featureNames(eset.main) ,c("GeneSymbol", "EntrezGeneID", "GeneTitle")]


```

# Get ENSEMBL annotation using biomaRt

```{r}

library(biomaRt)

mart <- useMart("ensembl")

mart <- useMart("ensembl", dataset="mmusculus_gene_ensembl")

attr <- listAttributes(mart)

attr[grep("affy", attr$name),]

genes <- getBM(attributes = c("ensembl_gene_id","external_gene_name", "description","affy_mogene_2_1_st_v1"), filters="affy_mogene_2_1_st_v1", values=featureNames(eset.main), mart=mart)

### clean the description
genes$description <- strsplit2(genes$description, " \\[Source")[, 1]

### Merge the info about multiple genes into one string
library(plyr)

genes.m <- ddply(genes, "affy_mogene_2_1_st_v1", summarize, GeneSymbol_Ensembl = paste0(external_gene_name, collapse = " /// "), GeneTitle_Ensembl = paste0(description, collapse = " /// "), EnsemblGeneID = paste0(ensembl_gene_id, collapse = " /// "))

annot.m <- merge(annot, genes.m, by.x = 0, by.y = "affy_mogene_2_1_st_v1", all.x = TRUE, sort = FALSE)
colnames(annot.m)[1] <- "ProbesetID"
rownames(annot.m) <- annot.m[,"ProbesetID"]

annot.m[is.na(annot.m)] <- "---"
all(annot.m$ProbesetID == featureNames(eset.main))

fData(eset.main) <- annot.m[featureNames(eset.main), ]
eset.main.org <- eset.main
```


# Filtering probe sets with low expression 

Filter probesets with low expression. For that use the background information from antigenomic probesets and GC content.

```{r}
### Get probe 2 transcript match
probeInfo <- oligo::getProbeInfo(x, field = c('fid', 'fsetid', 'level', 'type', 'transcript_cluster_id'), probeType = "pm", target='core')

pmSeq <- oligo::pmSequence(x, target='core')

### Get GC content per probe
library(Biostrings)
gcCont <- letterFrequency(pmSeq, letters='CG')[,1]

probeInfo$gcCont <- gcCont

### Get the background expression levels for different GC ammount
antigm <- infoNetAffx[infoNetAffx$category == "control->bgp->antigenomic", "probesetid"]

bgExpr <- exprs(eset)[as.character(antigm), targets$labels != "control_HeLa"]

bgProbeInfo <- subset(probeInfo, probeInfo$type == "control->bgp->antigenomic")

### see how many probes are for each GC content
table(bgProbeInfo$gcCont)


library(plyr)
library(ggplot2)
library(reshape2)

bgTransInfo <- ddply(bgProbeInfo, "transcript_cluster_id", summarize,  gcCont=mean(gcCont))

expr <- exprs(eset.main)

### Get the GC content for all the probe sets
transInfo <- ddply(probeInfo, "transcript_cluster_id", summarize,  gcCont=mean(gcCont))
rownames(transInfo) <- transInfo$transcript_cluster_id
transInfo <- transInfo[rownames(expr), ]
transInfo$gcCont <- round(transInfo$gcCont)
### see what is the average GC content for main probe sets
table(transInfo$gcCont)

### plot expression for background and main probes across different GC content
bgdf <- data.frame(bgTransInfo, bgExpr)

bgdf.m <- melt(bgdf, id.vars = c("transcript_cluster_id", "gcCont"), variable.name = "Samples", value.name = "Expression")

df <- data.frame(transInfo, expr)

df.m <- melt(df, id.vars = c("transcript_cluster_id", "gcCont"), variable.name = "Samples", value.name = "Expression")

df.m$Type <- "Main"
bgdf.m$Type <- "BGP"

df.all <- rbind(df.m, bgdf.m)
df.all$gcCont <- factor(df.all$gcCont, levels = 3:25)

ggp <- ggplot(data = df.all, aes(x = gcCont, y = Expression, fill = Type)) +
  geom_boxplot() +
  theme_bw() +
  theme(legend.position="top")

print(ggp)


################### set the threshold for each probe set
library(matrixStats)

bgTransInfo$Q095Expr <- rowQuantiles(bgExpr, probs = 0.95)
bgTransInfo

transInfo$minExpr <- factor(transInfo$gcCont, levels = bgTransInfo$gcCont)
levels(transInfo$minExpr) <- bgTransInfo$Q095Expr
transInfo$minExpr <- as.numeric(as.character(transInfo$minExpr))

```


# Comparison 1: control VS leukemia

Compare leukemia samples versus 3 different controls: CD4+, CD4+CD8+ and CD8+ and bone marrow.

```{r}
targets <- targets.org
eset.main <- eset.main.org

### keep only leukemia and control CD4+, CD4+CD8+ and CD8+ and bone marrow samples
samples2keep <- grepl("leukemia|control", targets$labels)

targets <- targets[samples2keep,]
eset.main <- eset.main[, samples2keep]

### sort samples by groups
ord <- order(targets$groups)
targets <- targets[ord, ]
eset.main <- eset.main[ ,ord]

# all(sampleNames(eset.main) == strsplit2(targets$FileName, "//")[,2])

#### Filtering itself
expr <- exprs(eset.main)
all(rownames(expr) == transInfo$transcript_cluster_id)

keepEXPR <- sapply(1:nrow(expr), function(tr){ sum(expr[tr, ] > transInfo$minExpr[tr]) >= 3 } )

table(keepEXPR)


eset.main <- eset.main[keepEXPR, ]

eset.main


#### design & analysis

treatments <- data.frame(Treatment = as.character(targets$groups))
treatments

design <- model.matrix(~ 0 + Treatment, data=treatments)
rownames(design) <- targets$labels
design

fit <- lmFit(eset.main, design)

contrasts <- cbind(CtrlCD4 = c(-1, 0, 0, 0, 1), CtrlCD4CD8 = c(0, -1, 0, 0, 1), CtrlCD8 = c(0, 0, -1, 0, 1), CtrlBM = c(0, 0, 0, -1, 1)) # treatment - control
contrasts


fit2 <- contrasts.fit(fit, contrasts)

fit2 <- eBayes(fit2, trend = TRUE)


plotSA(fit2)


results <- decideTests(fit2, method="separate", adjust.method="BH", p.value=0.05, lfc=0)
summary(results)

## with the FC cutoff
results <- decideTests(fit2, method="separate", adjust.method="BH", p.value=0.05, lfc=1)
summary(results)


colours <- unique(targets[targets$groups != "leukemia", "colors"])


vennDiagram(results,include=c("up", "down"), circle.col=colours, counts.col=colours)



### save all results with better order
coefs <- c("CtrlCD4", "CtrlCD4CD8", "CtrlCD8", "CtrlBM")

resExpr <- round(exprs(eset.main.org[keepEXPR, ]), 2)
colnames(resExpr) <- paste0(targets.org$groups, "_", colnames(resExpr))

resCoeff <- fit2$coefficients
colnames(resCoeff) <- paste0(colnames(resCoeff), "_coeffs")
resT <- fit2$t
colnames(resT) <- paste0(colnames(resT), "_t")
resPValue <- fit2$p.value
colnames(resPValue) <- paste0(colnames(resPValue), "_PValues")
resPValueAdj <- apply(fit2$p.value, 2, p.adjust, method = "BH")
colnames(resPValueAdj) <- paste0(colnames(resPValueAdj), "_AdjPValues")
resGenes <- fit2$genes
resRes <- results
colnames(resRes) <- paste0(colnames(resRes), "_Results")

stats <- c("coeffs", "t", "PValues", "AdjPValues", "Results")
colOrder <- paste(rep(coefs, each = length(stats)), rep(stats, length(coefs)), sep = "_")


resDE <- data.frame(resCoeff, resT, resPValue, resPValueAdj, resRes)[, colOrder]

resAll <- cbind(resGenes, resDE, resExpr[, order(colnames(resExpr))] )

write.table(resAll, file = "Comp1_DE_results_All.xls", quote = FALSE, sep = "\t", row.names = FALSE)



```


```{r, echo = FALSE}
# Multiple plot function
#
# ggplot objects can be passed in ..., or to plotlist (as a list of ggplot objects)
# - cols:   Number of columns in layout
# - layout: A matrix specifying the layout. If present, 'cols' is ignored.
#
# If the layout is something like matrix(c(1,2,3,3), nrow=2, byrow=TRUE),
# then plot 1 will go in the upper left, 2 will go in the upper right, and
# 3 will go all the way across the bottom.
#
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  require(grid)

  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)

  numPlots = length(plots)

  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                    ncol = cols, nrow = ceiling(numPlots/cols))
  }

 if (numPlots==1) {
    print(plots[[1]])

  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))

    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))

      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}

```




```{r, fig.width = 11, fig.height = 11}
par(mfrow=c(2,2))

### plot MA
for(i in 1:length(coefs)){
  coef <- coefs[i]
  limma::plotMA(fit2, coef = coef, status = results[, coef], values = c(-1, 0, 1), col = c("red", "black", "green"), cex = c(0.7, 0.3, 0.7), main = coef)
  abline(0,0,col="blue")
}

### volcano plots
library(ggplot2)


coefs <- c("CtrlCD4", "CtrlCD4CD8", "CtrlCD8", "CtrlBM")

gg1 <- list()
for(i in 1:length(coefs)){
coef <- coefs[i] 
table <- topTable(fit2, coef=coef, n=Inf)
table$threshold = as.factor(table$adj.P.Val < 0.05 & abs(table$logFC) > 1)
gg1[[i]] <- ggplot(data=table, aes(x=logFC, y=-log10(P.Value), colour=threshold)) +
  geom_point(alpha=0.4, size=1.75) + theme_bw() +ggtitle(coef) +
  theme(legend.position = "none") +
  xlab("log2 fold change") + ylab("-log10 p-value")

}

print(multiplot(plotlist = gg1, cols=2))

### histograms of p-values and adjusted p-values
colours <- unique(targets[targets$groups != "leukemia", "colors"])

for(i in 1:length(coefs)){
  
  coef <- coefs[i]  
  table <- topTable(fit2, coef=coef, n=Inf)
  hist(table$P.Value, breaks = 100, main = coef, xlab = "P-values", col = colours[i])

  
}

par(mfrow=c(1,1))
```


Plot expression of top significant genes/probe sets.

```{r, fig.width = 11, fig.height = 11}

library(ggplot2)
library(reshape2)

topn <- 20
expr <- exprs(eset.main)
xs <- 1:ncol(expr)

for(i in 1:length(coefs)){
  # i = 1
  coef <- coefs[i]
  print(coef)
  
  tt <- topTable(fit2, coef=coef, n=Inf, p.value=0.05, lfc=1)

  ### in the report display only first gene symbol
  GeneSymbol <- strsplit2(head(tt[,"GeneSymbol"], topn), " /// ")[,1]
  GeneTitle <- paste0(substr(strsplit2(head(tt[,"GeneTitle"], topn), " /// ")[,1], 1, 30))
  
  print(data.frame(GeneSymbol = GeneSymbol, GeneTitle = GeneTitle , head(tt[, c("logFC", "AveExpr", "P.Value", "adj.P.Val")], topn)))
  
  topp <- rownames(tt)[1:topn]

  
  df <- data.frame(Gene = topp, expr[topp,])
  df.m <- reshape2::melt(df, id.vars = "Gene", value.name = "Expression", variable.name = "Sample")
  ### keep order of genes as in tt
  df.m$Gene <- factor(df.m$Gene, levels = topp)
  ### add Entrez ID to the facet labels
  lab.fct <- paste0(topp, "\n", strsplit2(tt[topp, "GeneSymbol"], " /// ")[,1])
  levels(df.m$Gene) <- lab.fct
  
  ggp <- ggplot(df.m, aes(x = Sample, y = Expression)) +  
    theme_bw() +
    theme(axis.text.x = element_text(angle = 80, hjust = 1, size = 10), plot.title = element_text(size = 16), strip.text.x = element_text(size = 10)) +
    scale_x_discrete(labels=targets$groups) +
    labs(title = coef, y = "Log2 expression") +
    geom_bar(stat = "identity", colour = targets$colors, fill = targets$colors) +
    facet_wrap(~ Gene, scales="free_y", ncol=4) 
  

  print(ggp)    

  
}


```

## Gene set enrichment analysis with C5 - GO genes sets

```{r}

# gene sets from MSigDB with ENTREZ IDs
load("MSigDB_v4_0/mouse_c5_v4.rdata")

mysets <- Mm.c5
length(mysets)

### keep the sets of interest
intrset <- read.table("Gene_Sets/Interesting_gene_sets_C5.txt", header = FALSE, sep = ",")[, 1]
intrset

intrset <- gsub("-", " ", intrset)
intrset <- gsub(" ", "_", intrset)

intrset <- toupper(intrset)
length(intrset)

sum(names(mysets) %in% intrset)

mysets <- mysets[intrset]



### Create an Index for camera
annot <- fData(eset.main)

EntrezGeneID <- strsplit2(annot$EntrezGeneID, " /// ")

nrow = nrow(EntrezGeneID)
ncol = ncol(EntrezGeneID)

Index <- lapply(mysets, function(ms){  
  eg <- matrix(EntrezGeneID %in% ms, nrow = nrow, ncol = ncol, byrow = FALSE)
  rowSums(eg) > 0 
})


### Add index information to the final results
IndexMx <- do.call(cbind, Index)
class(IndexMx) <- "numeric"
colnames(IndexMx) <- names(mysets)
IndexMx <- data.frame(ProbesetID = annot$ProbesetID, IndexMx)

resAll <- merge(resAll, IndexMx, by = "ProbesetID", sort = FALSE)

write.table(resAll, file = "Comp1_DE_results_AllPlus.xls", quote = FALSE, sep = "\t", row.names = FALSE)


#### design & analysis

treatments <- data.frame(Treatment = as.character(targets$groups))

design <- model.matrix(~ 0 + Treatment, data=treatments)
rownames(design) <- targets$labels
design

contrasts <- cbind(CtrlCD4 = c(-1, 0, 0, 0, 1), CtrlCD4CD8 = c(0, -1, 0, 0, 1), CtrlCD8 = c(0, 0, -1, 0, 1), CtrlBM = c(0, 0, 0, -1, 1)) # treatment - control
contrasts




### run CAMERA

gsea <- list()

coef <- "CtrlCD4"
gsea.tmp <- gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=FALSE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("NGenes","Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), NGenes = gsea[[coef]][,1], gsea[[coef]][,-1])


### Example of a plot to show the profiles of genes in a gene set
topgs <- 1
gsn <-rownames(gsea[[coef]])[1:topgs] 
gss <- gsea.tmp[gsn, , drop = FALSE]

i = 1

  barcodeplot(statistics = as.numeric((fit2$t[,coef])), index = Index[[gsn[i]]], index2 = NULL, gene.weights = as.numeric((fit2$coefficients[, coef]))[Index[[gsn[i]]]], weights.label = "logFC", labels = c("Up","Down"), quantiles = c(-1,1), col.bars = NULL, worm = TRUE, span.worm=0.45, main = paste0(gsn[i], "\n", gss[i, "Direction"], ", FDR = ", sprintf("%.02e",gss[i, "FDR"])))
  
  barcodeplot(statistics = as.numeric((fit2$p.value[, coef])), index = Index[[gsn[i]]], index2 = NULL, gene.weights = as.numeric((fit2$coefficients[, coef]))[Index[[gsn[i]]]], weights.label = "logFC", labels = c("Not significant","Significant"), quantiles = c(0.05,1), col.bars = NULL, worm = TRUE, span.worm=0.45, main = paste0(gsn[i], "\n", gss[i, "Direction"], ", FDR = ", sprintf("%.02e",gss[i, "FDR"])))
  
  



coef <- "CtrlCD4CD8"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


coef <- "CtrlCD8"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


coef <- "CtrlBM"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


### merge all results into one table
gseaAll <- merge(gsea[["CtrlCD4"]], gsea[["CtrlCD4CD8"]], by = "GeneSet", all = TRUE)
gseaAll <- merge(gseaAll, gsea[["CtrlCD8"]], by = "GeneSet", all = TRUE)
gseaAll <- merge(gseaAll, gsea[["CtrlBM"]], by = "GeneSet", all = TRUE)
write.table(gseaAll, paste("Comp1_GSEA_C5_All.xls", sep=""), sep="\t", row.names=F, quote = FALSE)


```



## Gene set enrichment analysis with C7 - Immunologic genes sets

```{r gene sets}
# gene sets from MSigDB with ENTREZ IDs
load("MSigDB_v4_0/mouse_c7_v4.rdata")

mysets <- Mm.c7
length(mysets)


### Create an Index for camera
annot <- fData(eset.main)

EntrezGeneID <- strsplit2(annot$EntrezGeneID, " /// ")

nrow = nrow(EntrezGeneID)
ncol = ncol(EntrezGeneID)

Index <- lapply(mysets, function(ms){  
  eg <- matrix(EntrezGeneID %in% ms, nrow = nrow, ncol = ncol, byrow = FALSE)
  rowSums(eg) > 0 
})


### run CAMERA

gsea <- list()

coef <- "CtrlCD4"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=FALSE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("NGenes","Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), NGenes = gsea[[coef]][,1], gsea[[coef]][,-1])

coef <- "CtrlCD4CD8"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


coef <- "CtrlCD8"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


coef <- "CtrlBM"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


### merge all results into one table
gseaAll <- merge(gsea[["CtrlCD4"]], gsea[["CtrlCD4CD8"]], by = "GeneSet", all = TRUE)
gseaAll <- merge(gseaAll, gsea[["CtrlCD8"]], by = "GeneSet", all = TRUE)
gseaAll <- merge(gseaAll, gsea[["CtrlBM"]], by = "GeneSet", all = TRUE)
write.table(gseaAll, paste("Comp1_GSEA_C7_All.xls", sep=""), sep="\t", row.names=F, quote = FALSE)


```

## Gene set enrichment analysis with Hallmark genes sets

```{r}
############### Create mouse_hallmark_v5.rdata object like on WEHI web

allLines <- readLines("MSigDB_v4_0/h.all.v5.0.entrez.gmt", n = -1)

humanSets <- data.frame(strsplit2(allLines, "\t"), stringsAsFactors = FALSE)

namesHS <- humanSets[, 1]

Hu.hallmark <- apply(humanSets, 1, function(r){ 
  r <- r[-c(1,2)]
  r <- r[r != ""]
  return(as.numeric(r))
  } )


names(Hu.hallmark) <- namesHS

### get the mouse human homology
hom <- read.table("MSigDB_v4_0/HOM_MouseHumanSequence.txt", header = TRUE, sep = "\t")

homM <- hom[hom$Common.Organism.Name == "mouse, laboratory", c("HomoloGene.ID", "EntrezGene.ID")]

homH <- hom[hom$Common.Organism.Name == "human", c("HomoloGene.ID", "EntrezGene.ID")]

homMatch <- merge(homH, homM, by = "HomoloGene.ID", sort = FALSE, all = TRUE) 

homMatch <- homMatch[!is.na(homMatch[, 2]) & !is.na(homMatch[, 3]), ]


Mm.hallmark <- lapply(Hu.hallmark, function(gs){
  
  unique(homMatch[homMatch[, 2] %in% gs, 3])
  
})


save(Mm.hallmark, file = "MSigDB_v4_0/mouse_hallmark_v5.rdata")


```


```{r}

# gene sets from MSigDB with ENTREZ IDs
load("MSigDB_v4_0/mouse_hallmark_v5.rdata")

mysets <- Mm.hallmark
length(mysets)

### keep the sets of interest
intrset <- read.table("Gene_Sets/Interesting_gene_sets_Hallmark.txt", header = FALSE, sep = ",", as.is = TRUE)[, 1]
intrset

intrset <- gsub("-", " ", intrset)
intrset <- gsub(" ", "_", intrset)

intrset <- paste0("HALLMARK_",toupper(intrset))
length(intrset)

sum(names(mysets) %in% intrset)

mysets <- mysets[intrset]

### Create an Index for camera
annot <- fData(eset.main)


EntrezGeneID <- strsplit2(annot$EntrezGeneID, " /// ")

nrow = nrow(EntrezGeneID)
ncol = ncol(EntrezGeneID)

Index <- lapply(mysets, function(ms){  
  eg <- matrix(EntrezGeneID %in% ms, nrow = nrow, ncol = ncol, byrow = FALSE)
  rowSums(eg) > 0 
})


IndexMx <- do.call(cbind, Index)
class(IndexMx) <- "numeric"
colnames(IndexMx) <- names(mysets)
IndexMx <- data.frame(ProbesetID = annot$ProbesetID, IndexMx)

resAll <- merge(resAll, IndexMx, by = "ProbesetID", sort = FALSE)

write.table(resAll, file = "Comp1_DE_results_AllPlus.xls", quote = FALSE, sep = "\t", row.names = FALSE)


#### design & analysis

treatments <- data.frame(Treatment = as.character(targets$groups))

design <- model.matrix(~ 0 + Treatment, data=treatments)
rownames(design) <- targets$labels
design

contrasts <- cbind(CtrlCD4 = c(-1, 0, 0, 0, 1), CtrlCD4CD8 = c(0, -1, 0, 0, 1), CtrlCD8 = c(0, 0, -1, 0, 1), CtrlBM = c(0, 0, 0, -1, 1)) # treatment - control
contrasts


### run CAMERA

gsea <- list()

coef <- "CtrlCD4"
gsea.tmp <- gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=FALSE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("NGenes","Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), NGenes = gsea[[coef]][,1], gsea[[coef]][,-1])
# write.table(gsea[[coef]], paste("Comp1_GSEA_",coef ,".xls", sep=""), sep="\t", row.names=F, quote = FALSE)



### using information from eBayes fitting: fit2


topgs <- 1
gsn <-rownames(gsea[[coef]])[1:topgs] 
gss <- gsea.tmp[gsn, , drop = FALSE]

i = 1

  barcodeplot(statistics = as.numeric((fit2$t[,coef])), index = Index[[gsn[i]]], index2 = NULL, gene.weights = as.numeric((fit2$coefficients[, coef]))[Index[[gsn[i]]]], weights.label = "logFC", labels = c("Up","Down"), quantiles = c(-1,1), col.bars = NULL, worm = TRUE, span.worm=0.45, main = paste0(gsn[i], "\n", gss[i, "Direction"], ", FDR = ", sprintf("%.02e",gss[i, "FDR"])))
  
  barcodeplot(statistics = as.numeric((fit2$p.value[, coef])), index = Index[[gsn[i]]], index2 = NULL, gene.weights = as.numeric((fit2$coefficients[, coef]))[Index[[gsn[i]]]], weights.label = "logFC", labels = c("Not significant","Significant"), quantiles = c(0.05,1), col.bars = NULL, worm = TRUE, span.worm=0.45, main = paste0(gsn[i], "\n", gss[i, "Direction"], ", FDR = ", sprintf("%.02e",gss[i, "FDR"])))
  
  




coef <- "CtrlCD4CD8"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


coef <- "CtrlCD8"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


coef <- "CtrlBM"
gsea[[coef]] <- camera(y = eset.main, index=Index, design=design, contrast=contrasts[,coef], trend.var=TRUE)
head(gsea[[coef]], 10)
table(gsea[[coef]]$FDR < 0.05)
gsea[[coef]] <- gsea[[coef]][, c("Direction", "PValue", "FDR")]
colnames(gsea[[coef]]) <- paste0(coef, "_", colnames(gsea[[coef]]))
gsea[[coef]] <- data.frame(GeneSet = rownames(gsea[[coef]]), gsea[[coef]])


### merge all results into one table
gseaAll <- merge(gsea[["CtrlCD4"]], gsea[["CtrlCD4CD8"]], by = "GeneSet", all = TRUE)
gseaAll <- merge(gseaAll, gsea[["CtrlCD8"]], by = "GeneSet", all = TRUE)
gseaAll <- merge(gseaAll, gsea[["CtrlBM"]], by = "GeneSet", all = TRUE)
write.table(gseaAll, paste("Comp1_GSEA_Hallmark_All.xls", sep=""), sep="\t", row.names=F, quote = FALSE)



```


## GO analysis

```{r GO, results = "hide", eval = FALSE}

# results = "hide"
# include=FALSE

library(topGO)
library(Rgraphviz)


affyLib <- "mogene20sttranscriptcluster.db"

### Function used to create new topGOdata object
fun.gene.sel <- function(geneList) {
  return(geneList <- ifelse(geneList==0, FALSE, TRUE))
}

coefs <- c("CtrlCD4", "CtrlCD4CD8", "CtrlCD8", "CtrlBM")

allResList <- list()

for(coef in coefs){
  # coef <- coefs[1]
  
  tt <- topTable(fit2, coef=coef, n=Inf)
  geneList <- rep(0, nrow(tt))
  names(geneList) <- rownames(tt)
  geneList[tt$adj.P.Val < 0.05 & abs(tt$logFC) > 1] <- 1
  print(table(geneList))
    
  for(go in c("BP","MF","CC")){
    # go = "BP"
    print(coef)
    print(go)
    sampleGOdata <- new("topGOdata", description = paste0("Simple session for ", coef), ontology = go, allGenes = geneList, geneSel = fun.gene.sel , nodeSize = 10, annot = annFUN.db, affyLib = affyLib)
    
    print(sampleGOdata)
    
    result <- runTest(sampleGOdata, algorithm = "elim", statistic = "fisher")
    
    pValues <- score(result)
    topNodes <- length(pValues)
    
    allRes <- GenTable(sampleGOdata, elimFisher = result, orderBy = "elimFisher", topNodes = topNodes)      
    colnames(allRes)[6] <- "PValues" 
     
    allRes$AdjPValues <- p.adjust(allRes$PValues, method = "BH")
    
    print(table(allRes$AdjPValues < 0.05))
    
    cat("#########################################################################################", fill = TRUE)
    print(head(allRes, 20))
    cat("#########################################################################################", fill = TRUE)
    
 
    allResList[[paste0(go, "_", coef)]] <- allRes
    
    
  }
  
}


#### save results 

for(go in c("BP","MF","CC")){

  coef <- "CtrlCD4"
  allR <- allResList[[paste0(go, "_", coef)]]
  colnames(allR) <- paste0(c(rep("", 3), rep(paste0(coef, "_"), 4)), colnames(allR))
  allResList[[paste0(go, "_", coef)]] <- allR
  
  coef <- "CtrlCD4CD8"
  allR <- allResList[[paste0(go, "_", coef)]][, -c(2, 3)]
  colnames(allR) <- paste0(c("", rep(paste0(coef, "_"), 4)), colnames(allR))
  allResList[[paste0(go, "_", coef)]] <- allR
  
  
  coef <- "CtrlCD8"
  allR <- allResList[[paste0(go, "_", coef)]][, -c(2, 3)]
  colnames(allR) <- paste0(c("", rep(paste0(coef, "_"), 4)), colnames(allR))
  allResList[[paste0(go, "_", coef)]] <- allR
  

  coef <- "CtrlBM"
  allR <- allResList[[paste0(go, "_", coef)]][, -c(2, 3)]
  colnames(allR) <- paste0(c("", rep(paste0(coef, "_"), 4)), colnames(allR))
  allResList[[paste0(go, "_", coef)]] <- allR
  
  
  ### merge all results into one table
  allAll <- merge(allResList[[paste0(go, "_", "CtrlCD4")]], allResList[[paste0(go, "_", "CtrlCD4CD8")]], by = "GO.ID", all = TRUE)  
  allAll <- merge(allAll, allResList[[paste0(go, "_", "CtrlCD8")]], by = "GO.ID", all = TRUE)
  allAll <- merge(allAll, allResList[[paste0(go, "_", "CtrlBM")]], by = "GO.ID", all = TRUE)
  write.table(allAll, paste0("Comp1_GO_Fisher_elim_", go ,".xls"), sep="\t", row.names=F, quote = FALSE)
    
}



```

# Comparison 2: pre VS after treatment

## All leukemia VS afterTreatment samples

```{r}
targets <- targets.org
eset.main <- eset.main.org

### keep only leukemia and afterTreatment samples
samples2keep <- grepl("leukemia|afterTreatment", targets$labels)

targets <- targets[samples2keep,]
eset.main <- eset.main[, samples2keep]

### sort samples by groups
ord <- order(targets$groups)
targets <- targets[ord, ]
eset.main <- eset.main[ ,ord]

# all(sampleNames(eset.main) == strsplit2(targets$FileName, "//")[,2])

#### Filtering itself
expr <- exprs(eset.main)
all(rownames(expr) == transInfo$transcript_cluster_id)

keepEXPR <- sapply(1:nrow(expr), function(tr){ sum(expr[tr, ] > transInfo$minExpr[tr]) >= 2 } )

table(keepEXPR)


eset.main <- eset.main[keepEXPR, ]

eset.main




### plot MDS
pdf("PLOTS/MDS_norm_Comp2.pdf")
mds <- plotMDS(eset.main, top=500, col = targets$colors, labels = targets$labels, cex = 1.2, xlim = c(-2, 3), ylim = c(-2, 3))
dev.off()




###########################################################################

# gene sets from MSigDB with ENTREZ IDs / C6 - oncogenic signatures
load("MSigDB_v4_0/mouse_c6_v4.rdata")

# Mm.c6[1]
mysets <- Mm.c6
length(mysets)

# table(sapply(mysets, length))


### Create an Index for camera
annot <- fData(eset.main)
# table(annot$EntrezGeneID == "---")

EntrezGeneID <- strsplit2(annot$EntrezGeneID, " /// ")

nrow = nrow(EntrezGeneID)
ncol = ncol(EntrezGeneID)

Index <- lapply(mysets, function(ms){  
  eg <- matrix(EntrezGeneID %in% ms, nrow = nrow, ncol = ncol, byrow = FALSE)
  rowSums(eg) > 0 
})



###########################################################################


```


### Account for the Cell Type

```{r}
#### design & analysis

treatments <- data.frame(Treatment = as.character(targets$groups), CellType = as.character(targets$CellTypeShort))

treatments$Treatment <- factor(treatments$Treatment, levels = c("leukemia", "afterTreatment")) 

design <- model.matrix(~ 0 + CellType + Treatment, data=treatments)
rownames(design) <- targets$labels
design

fit <- lmFit(eset.main, design)

fit2 <- eBayes(fit[, "TreatmentafterTreatment"], trend = TRUE)


## with the FC cutoff
results <- decideTests(fit2, method="separate", adjust.method="BH", p.value=0.05, lfc=1)
summary(results)


table <- topTable(fit2, coef=1, n=Inf)
### in the report display only first gene symbol
topn <- 10
GeneSymbol <- strsplit2(head(table[,"GeneSymbol"], topn), " /// ")[,1]
GeneTitle <- paste0(substr(strsplit2(head(table[,"GeneTitle"], topn), " /// ")[,1], 1, 30))  
print(data.frame(GeneSymbol = GeneSymbol, GeneTitle = GeneTitle , head(table[, c("logFC", "AveExpr", "P.Value", "adj.P.Val")], topn)))

hist(table$P.Value, breaks = 100, xlab = "P-values")  



#### Gene set test CAMERA
gsea <- camera(y = eset.main, index=Index, design=design, contrast=ncol(design), trend.var=TRUE)
head(gsea, 10)
table(gsea$FDR < 0.05)


```

### Do not account for the Cell Type

```{r}
#### design & analysis

treatments <- data.frame(Treatment = as.character(targets$groups), CellType = as.character(targets$CellTypeShort))

treatments$Treatment <- factor(treatments$Treatment, levels = c("leukemia", "afterTreatment")) 

design <- model.matrix(~ Treatment, data=treatments)
rownames(design) <- targets$labels
design

fit <- lmFit(eset.main, design)

fit2 <- eBayes(fit[, "TreatmentafterTreatment"], trend = TRUE)


## with the FC cutoff
results <- decideTests(fit2, method="separate", adjust.method="BH", p.value=0.05, lfc=1)
summary(results)


table <- topTable(fit2, coef=1, n=Inf)
### in the report display only first gene symbol
topn <- 10
GeneSymbol <- strsplit2(head(table[,"GeneSymbol"], topn), " /// ")[,1]
GeneTitle <- paste0(substr(strsplit2(head(table[,"GeneTitle"], topn), " /// ")[,1], 1, 30))  
print(data.frame(GeneSymbol = GeneSymbol, GeneTitle = GeneTitle , head(table[, c("logFC", "AveExpr", "P.Value", "adj.P.Val")], topn)))

hist(table$P.Value, breaks = 100, xlab = "P-values")  




#### Gene set test CAMERA
gsea <- camera(y = eset.main, index=Index, design=design, contrast=ncol(design), trend.var=TRUE)
head(gsea, 10)
table(gsea$FDR < 0.05)


```



## 907 and B2M3 leukemia VS afterTreatment samples

```{r}
targets <- targets.org
eset.main <- eset.main.org

### keep only leukemia and afterTreatment samples
samples2keep <- grepl("leukemia|afterTreatment", targets$labels) & grepl("907|B2M3", targets$CellTypeShort)

targets <- targets[samples2keep,]
eset.main <- eset.main[, samples2keep]

### sort samples by groups
ord <- order(targets$groups)
targets <- targets[ord, ]
eset.main <- eset.main[ ,ord]

# all(sampleNames(eset.main) == strsplit2(targets$FileName, "//")[,2])

#### Filtering itself
expr <- exprs(eset.main)
all(rownames(expr) == transInfo$transcript_cluster_id)

keepEXPR <- sapply(1:nrow(expr), function(tr){ sum(expr[tr, ] > transInfo$minExpr[tr]) >= 2 } )

table(keepEXPR)


eset.main <- eset.main[keepEXPR, ]

eset.main


###########################################################################

# gene sets from MSigDB with ENTREZ IDs / C6 - oncogenic signatures
load("MSigDB_v4_0/mouse_c6_v4.rdata")

# Mm.c6[1]
mysets <- Mm.c6
length(mysets)

# table(sapply(mysets, length))


### Create an Index for camera
annot <- fData(eset.main)
# table(annot$EntrezGeneID == "---")

EntrezGeneID <- strsplit2(annot$EntrezGeneID, " /// ")

nrow = nrow(EntrezGeneID)
ncol = ncol(EntrezGeneID)

Index <- lapply(mysets, function(ms){  
  eg <- matrix(EntrezGeneID %in% ms, nrow = nrow, ncol = ncol, byrow = FALSE)
  rowSums(eg) > 0 
})

```

### Account for the Cell Type

```{r}
#### design & analysis

treatments <- data.frame(Treatment = as.character(targets$groups), CellType = as.character(targets$CellTypeShort))

treatments$Treatment <- factor(treatments$Treatment, levels = c("leukemia", "afterTreatment")) 

design <- model.matrix(~ 0 + CellType + Treatment, data=treatments)
rownames(design) <- targets$labels
design

fit <- lmFit(eset.main, design)

fit2 <- eBayes(fit[, "TreatmentafterTreatment"], trend = TRUE)


## with the FC cutoff
results <- decideTests(fit2, method="separate", adjust.method="BH", p.value=0.05, lfc=1)
summary(results)



table <- topTable(fit2, coef=1, n=Inf)
### in the report display only first gene symbol
topn <- 10
GeneSymbol <- strsplit2(head(table[,"GeneSymbol"], topn), " /// ")[,1]
GeneTitle <- paste0(substr(strsplit2(head(table[,"GeneTitle"], topn), " /// ")[,1], 1, 30))  
print(data.frame(GeneSymbol = GeneSymbol, GeneTitle = GeneTitle , head(table[, c("logFC", "AveExpr", "P.Value", "adj.P.Val")], topn)))

hist(table$P.Value, breaks = 100, xlab = "P-values")  




#### Gene set test CAMERA
gsea <- camera(y = eset.main, index=Index, design=design, contrast=ncol(design), trend.var=TRUE)
head(gsea, 10)
table(gsea$FDR < 0.05)



```

### Do not account for the Cell Type

```{r}

#### design & analysis

treatments <- data.frame(Treatment = as.character(targets$groups), CellType = as.character(targets$CellTypeShort))

treatments$Treatment <- factor(treatments$Treatment, levels = c("leukemia", "afterTreatment")) 

design <- model.matrix(~ Treatment, data=treatments)
rownames(design) <- targets$labels
design

fit <- lmFit(eset.main, design)

fit2 <- eBayes(fit[, "TreatmentafterTreatment"], trend = TRUE)


## with the FC cutoff
results <- decideTests(fit2, method="separate", adjust.method="BH", p.value=0.05, lfc=1)
summary(results)


table <- topTable(fit2, coef=1, n=Inf)
### in the report display only first gene symbol
topn <- 10
GeneSymbol <- strsplit2(head(table[,"GeneSymbol"], topn), " /// ")[,1]
GeneTitle <- paste0(substr(strsplit2(head(table[,"GeneTitle"], topn), " /// ")[,1], 1, 30))  
print(data.frame(GeneSymbol = GeneSymbol, GeneTitle = GeneTitle , head(table[, c("logFC", "AveExpr", "P.Value", "adj.P.Val")], topn)))

hist(table$P.Value, breaks = 100, xlab = "P-values")  


#### Gene set test CAMERA
gsea <- camera(y = eset.main, index=Index, design=design, contrast=ncol(design), trend.var=TRUE)
head(gsea, 10)
table(gsea$FDR < 0.05)

```





```{r}
sessionInfo()
```








